src/server/pfs/server/driver.go

```diff
@@ -1118,139 +1124,88 @@ func (d *driver) listCommit(ctx context.Context, repo *pfs.Repo, to *pfs.Commit,
+		return err
 	}
-
-	stream := make(chan CommitEvent)
-	done := make(chan struct{})
-
-	go func() (retErr error) {
-		defer newCommitWatcher.Close()
-		defer func() {
-			if retErr != nil {
-				select {
-				case stream <- CommitEvent{
-					Err: retErr,
-				}:
-				case <-done:
-				}
-			}
-			close(stream)
-		}()
-		// keep track of the commits that have been sent
-		seen := make(map[string]bool)
-		// include all commits that are currently on the given branch,
-		// but only the ones that have been finished
-		commitInfos, err := d.listCommit(ctx, repo, &pfs.Commit{
-			Repo: repo,
-			ID:   branch,
-		}, from, 0)
+	defer newCommitWatcher.Close()
+	// keep track of the commits that have been sent
+	seen := make(map[string]bool)
+	// include all commits that are currently on the given branch,
+	commitInfos, err := d.listCommit(ctx, repo, client.NewCommit(repo.Name, branch), from, 0)
+	if err != nil {
+		// We skip NotFound error because it's ok if the branch
+		// doesn't exist yet, in which case ListCommit returns
+		// a NotFound error.
+		if !isNotFoundErr(err) {
+			return err
+		}
+	}
+	// ListCommit returns commits in newest-first order,
+	// but SubscribeCommit should return commit in oldest-first
+	// order, so we reverse the order.
+	for i := range commitInfos {
+		commitInfo := commitInfos[len(commitInfos)-i-1]
+		commitInfo, err := d.inspectCommit(ctx, commitInfo.Commit, state)
 		if err != nil {
-			// We skip NotFound error because it's ok if the branch
-			// doesn't exist yet, in which case ListCommit returns
-			// a NotFound error.
-			if !isNotFoundErr(err) {
-				return err
-			}
+			return err
 		}
-		// ListCommit returns commits in newest-first order,
-		// but SubscribeCommit should return commit in oldest-first
-		// order, so we reverse the order.
-		for i := range commitInfos {
-			commitInfo := commitInfos[len(commitInfos)-i-1]
-			select {
-			case stream <- CommitEvent{
-				Value: commitInfo,
-			}:
-				seen[commitInfo.Commit.ID] = true
-			case <-done:
-				return nil
-			}
+		if err := f(commitInfo); err != nil {
+			return err
 		}
-
-		for {
-			var branchName string
-			branchInfo := &pfs.BranchInfo{}
-			var event *watch.Event
-			var ok bool
-			select {
-			case event, ok = <-newCommitWatcher.Watch():
-			case <-done:
-				return nil
-			}
-			if !ok {
-				return nil
-			}
-			switch event.Type {
-			case watch.EventError:
-				return event.Err
-			case watch.EventPut:
-				if err := event.Unmarshal(&branchName, branchInfo); err != nil {
-					return fmt.Errorf("Unmarshal: %v", err)
-				}
-			case watch.EventDelete:
-				continue
-			}
-			if branchInfo.Head == nil {
-				continue // put event == new branch was created. No commits yet though
+		seen[commitInfo.Commit.ID] = true
+	}
+	for {
+		var branchName string
+		branchInfo := &pfs.BranchInfo{}
+		var event *watch.Event
+		var ok bool
+		event, ok = <-newCommitWatcher.Watch()
+		if !ok {
+			return nil
+		}
+		switch event.Type {
```

Comment:
It would probably be more readable if the code following this switch was in the put event case because it would be clear upon first look that it corresponds to a put event. Also, you could get rid of the delete event case or leave it empty if you did this.

Reply:
Good call, will do.